<a href="https://colab.research.google.com/github/AndresVelez31/AlgebraLineal/blob/main/Gu%C3%ADa_de_Python_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# **Guía #1: Álgebra de Vectores**

**Objetivo:** Usar el lenguaje de programación de Python para el desarrollo y apropiación de las propiedades básicas de los vectores en $\mathbb{R}^2$ y $\mathbb{R}^3$.

**Temas:** En esta guía se trabajarán las distintas  operaciones entre vectores como:  suma, resta, producto escalar, producto punto, producto vectorial, norma, ortogonalidad, ángulo entre vectores y proyección ortogonal.




## **Creación de vectores**

Para la creación y tratado de vectores se debe llamar la biblioteca "**numpy**". Ésta biblioteca da soporte para crear vectores y matrices, y permite la realización de las diversas operaciones entre éstos elementos. "**Numpy**" generalmente se importa con el alias "**np**". En los siguientes ejemplos se muestra cómo llamar esta biblioteca.

### **Ejemplo 1:**
Crear los siguientes vectores:



\begin{equation*}
 \vec{u} = (1, 40, -20),
 \hspace{1cm}
 \vec{v} = (-9, 32, 61),
\hspace{1cm}
 \vec{w} = (0, -345, 42),
 \hspace{1cm}
 \vec{z} = (49, 2, 7),
\end{equation*}


In [ ]:
# Solución del Ejemplo 1.

# Importamos la biblioteca numpy con el alias np.
import numpy as np

# Creación de los vctores u, v, w, z.

u = np.array([1, 40, -20])

v = np.array([-9, 32, 61])

w = np.array([0, -345, 42])

z = np.array([49, 2, 7])

# Imprimo los vectores

print('u =', u), print('v =', v), print('w =', w), print('z =', z);

u = [  1  40 -20]
v = [-9 32 61]
w = [   0 -345   42]
z = [49  2  7]


## **Suma, resta y producto escalar**

### **Ejemplo 2:**
Use los vectores establecidos en el Ejemplo 1 para realizar las operaciones indicadas:



a).   $(2\vec{u} + 6 \vec{w}) -(7\vec{v}-\vec{z}).$

b).   $-2\vec{w} + (\vec{v}-4\vec{u}).$

c).  $4\vec{v} - (\vec{u}-3\vec{w}) + 3\vec{z}.$





In [ ]:
# Solución del Ejemplo 2.

# a). Operación (2u + 6w)-(7v-z)

a = (2*u + 6*w)-(7*v-z)

# Imprimo el vector a

print('a =', a);

a = [  114 -2212  -208]


In [ ]:
# b). Operación -2w + (v-4u)

b = -2*w + (v-4*u)

# Imprimo el vector b

print('b =', b);

b = [-13 562  57]


In [ ]:
# c). Operación 4v - (u-3w) + 3z

c = 4*v - (u-3*w) + 3*z

# Imprimo el vector c

print('c =', c);

c = [ 110 -941  411]


## **Producto punto y magnitud (norma)**

### **Ejemplo 3:**
Crear los siguientes vectores

\begin{equation*}
 \vec{u} = (50, -200, 0),
 \hspace{1cm}
 \vec{v} = (8, 609, 12),
\hspace{1cm}
 \vec{w} = (-4, -5, 2),
 \hspace{1cm}
 \vec{z} = (10, 24, 78),
\end{equation*}

y realizar las operaciones indicadas:

a). $ \vec{u}\cdot \vec{z} + \vec{w}\cdot (3\vec{v}).$

b). $ (\vec{u} + \vec{z}- \vec{v}) \cdot (\vec{z}-3\vec{w} + \vec{v}).$

c). $\| \vec{z}- \vec{w}\|- \|\vec{w} + \vec{u}\|.$

d). $\|\vec{u}\|(\vec{v} + \vec{z}) + \| 2\vec{z}- 3\vec{w}\|\vec{z}.$

In [ ]:
# Solución del Ejemplo 3.

# Creación de los vectores u, v, w, z.

u = np.array([50, -200, 0])

v = np.array([8, 609, 12])

w = np.array([-4, -5, 2])

z = np.array([10, 24, 78])

# Imprimo los vectores

print('u =', u), print('v =', v), print('w =', w), print('z =', z);

u = [  50 -200    0]
v = [  8 609  12]
w = [-4 -5  2]
z = [10 24 78]


In [ ]:
# a). Operación u.z + w.(3v)

# Nota: El producto punto entre vectores se puede ejecutar con el comando @

a = u @ z + w @ (3*v)

print('a =', a);

a = -13459


In [ ]:
# b). Operacion (u + z -v).(z - 3w + v)

b = (u + z -v) @ (z -3*w + v)

print('b =', b);

b = -501576


In [ ]:
# c). Operación ||z - w|| - ||w + u||

# Nota: Para calcular la norma de un vector podemos hacer uso del comando np.linalg.norm.

c = np.linalg.norm(z -w) - np.linalg.norm(w + u)

print('c =', c);

c = -127.56621658553189


In [ ]:
# d). Operación ||u||(v-z) + ||2z -3w||z

d = np.linalg.norm(u)*(v-z) + np.linalg.norm(2*z - 3*w)*z

print('d =', d);

d = [  1245.79076124 124580.28272645   -673.05823887]


## **Ángulo entre vectores**

**Ángulo entre vectores:** De las clases hemos aprendido que

$$ \cos \theta = \frac{\vec{u} \cdot \vec{v}}{\|\vec{u}\| \|\vec{v}\|},  $$

donde $\theta$ es el **ángulo que forman los vectores** $\vec{u}$ y $\vec{v}$. Por lo tanto, el ángulo deseado se puede estimar invirtiendo la función coseno.


### **Ejemplo 4:**
Determinar en cada caso el ángulo entre los vectores:


a).   $\vec{u} = (3,0), \ \vec{v} = (5, 5).$

b).   $ \vec{u} = (-1,3,4), \ \vec{v} = (5,-2,7).$




In [ ]:
# Solución ejemplo 4

# a). Defino los vectores u y v

u = np.array([3,0])

v = np.array([5,5])

# Calculo el producto punto de u con v

prod_uv = u @ v

# Calculo la norma de los vectores u y v

norm_u = np.linalg.norm(u)

norm_v = np.linalg.norm(v)

# Hallo el ángulo entre los vecotres u y v. Para ello, llamamos la biblioteca math de Python.

import math

# El inverso del coseno se obtiene mediante el comando math.acos

theta = math.acos(prod_uv/(norm_u*norm_v))

# Convierto el valor del ángulo en grados con el comando math.degrees

theta = math.degrees(theta)

print('theta =', theta);

theta = 45.00000000000001


De lo anterior, tenemos que el ángulo entre los vectores es de $\theta = 45^o$.

In [ ]:
# b). Defino los vectores u y v

u = np.array([-1,3,4])

v = np.array([5,-2,7])

# Calculo el producto punto de u con v

prod_uv = u @ v

# Calculo la norma de los vectores u y v

norm_u = np.linalg.norm(u)

norm_v = np.linalg.norm(v)

# El inverso del coseno se obtiene mediante el comando math.acos

theta = math.acos(prod_uv/(norm_u*norm_v))

# Convierto el valor del ángulo en grados con el comando math.degrees

theta = math.degrees(theta)

print('theta =', theta);

theta = 67.82119547323865


Luego, el ángulo entre los vectores es de $\theta \approx  67.8^o$.

## **Ortogonalidad y proyección**

**Ortogonalidad:** Como sabemos, dos vectores $\vec{u}$ y $\vec{v}$ son ortogonales si se cumple que **el producto punto entre ellos es cero**, es decir,

$$\vec{u} \cdot \vec{v} = 0.$$

### **Ejemplo 5:**
Determine si las parejas de vectores dadas a continuación son ortogonales:

a). $ \vec{u} = (1, 36,-8), \ \vec{v} = (9, -2,36)$

b).  $ \vec{u} = (14, 21,-7), \ \vec{v} = (3, -2,0)$


In [ ]:
# Solución ejemplo 5

# a).  Creo los vectores u y v

u = np.array([1,36, -8])

v = np.array([9, -2, 36])

# Calculo el producto punto entre u y v

prod_uv = u @ v

print('prod_uv =', prod_uv);

prod_uv = -351


Como $\vec{u} \cdot \vec{v} = -351$, entonces los vectores no son ortogonales.

In [ ]:
# b).  Creo los vectores u y v

u = np.array([14,21, -7])

v = np.array([3, -2, 0])

# Calculo el producto punto entre u y v

prod_uv = u @ v

print('prod_uv =', prod_uv);

prod_uv = 0


Como $\vec{u} \cdot \vec{v} = 0$, entonces los vectores son ortogonales.

**Proyección ortogonal:** Recordemos que dados los vectores $\vec{u}$ y $\vec{v}$, se define el vector proyección de $\vec{v}$ sobre $\vec{u}$ como:  

$$\vec{p} = Pr(\vec{v}/\vec{u}) = \frac{\vec{v} \cdot \vec{u}}{\|\vec{u}\|^2}\vec{u} = \lambda \vec{u},$$

donde $\lambda = \frac{\vec{v} \cdot \vec{u}}{\|\vec{u}\|^2} \in \mathbb{R}. $ Por lo tanto, el vector $\vec{p}$ es **paralelo** a $\vec{u}$ y el vector $\vec{h} = \vec{v}-\vec{p}$ es **ortogonal** a $\vec{u}.$

### **Ejemplo 6:**
Calcule la proyección de $\vec{v}$ sobre $\vec{u}$ en cada caso:

a). $ \vec{u} = (1, 6,8), \ \vec{v} = (8, -2,1)$

b).  $ \vec{u} = (1,0), \ \vec{v} = (3,-6)$


In [ ]:
# Solución ejemplo 6

# a).  Creo los vectores u y v

u = np.array([1,6,8])

v = np.array([8, -2, 1])

# Calculo la norma de u

norm_u = np.linalg.norm(u)

# Elevo al cuadrado la norma de u, para ello uso el comando math.pow

norm_u2 = math.pow(norm_u,2)

# Hallo el producto punto de u con v

prod_uv = u @ v

# Hallo la constante lambda

Lambda = prod_uv/norm_u2

# Hallo la proyección ortogonal p

p = Lambda*u

# visualizo el vector p

print('p =', p);

p = [0.03960396 0.23762376 0.31683168]


In [ ]:
# b).  Creo los vectores u y v

u = np.array([1,0])

v = np.array([3,-6])

# Calculo la norma de u

norm_u = np.linalg.norm(u)

# Elevo al cuadrado la norma de u, para ello uso el comando math.pow

norm_u2 = math.pow(norm_u,2) # aquí el número 2 que aparece corresponde a elevar al cuadrado

# Hallo el producto punto de u con v

prod_uv = u @ v

# Hallo la constante lambda

Lambda = prod_uv/norm_u2

# Hallo la proyección ortogonal p

p = Lambda*u

# visualizo el vector p

print('p =', p);

p = [3. 0.]


### **Ejemplo 7:**
Sean $\vec{u} = (2, 1, 6)$ y $\vec{v} = (1, −1, 3)$. Hallar dos vectores $\vec{p}$ y $\vec{h}$ tales que $\vec{v} = \vec{p} + \vec{h}$ con $\vec{p}$ paralelo a $\vec{u}$ y $\vec{h}$ ortogonal a $\vec{u}$.

**Solución:** De los ejemplos anteriores vemos que para encontrar estos vectores, podemos definir $\vec{p}$ como la proyección de $\vec{v}$ sobre $\vec{u}$, y finalmente, el vector $\vec{h}$ como $\vec{v}-\vec{p}.$

In [ ]:
# Solución ejemplo 7

# Creo los vectores u y v

u = np.array([2,1,6])

v = np.array([1,-1,3])

# Calculo la norma de u

norm_u = np.linalg.norm(u)

# Elevo al cuadrado la norma de u, para ello uso el comando math.pow

norm_u2 = math.pow(norm_u,2) # aquí el número 2 que aparece corresponde a elevar al cuadrado

# Hallo el producto punto de u con v

prod_uv = u @ v

# Hallo la constante lambda

Lambda = prod_uv/norm_u2

# Hallo la proyección ortogonal p

p = Lambda*u

# Defino el vector h = v-p

h = v-p

# Visualizo los vectores p y h

print('p =', p), print('h =', h);

p = [0.92682927 0.46341463 2.7804878 ]
h = [ 0.07317073 -1.46341463  0.2195122 ]


## **Producto vectorial (cruz) en $\mathbb{R}^3$**

### **Ejemplo 8:**
Halle el producto vectorial $\vec{u} \times \vec{v}$ en cada caso:

a). $\vec{u} = (1, 5,-45)$ y  $\vec{v} = (9, -2,0)$

b). $\vec{u} = (1, 0,0)$ y  $\vec{v} = (0, 1,0)$

In [ ]:
# Solución ejemplo 8

# a). Defino los vectores u y v

u = np.array([1,5,-45])

v = np.array([9,-2,0])

# Hallo el producto vectorial (producto cruz) de u y v, usando el comando np.cross

prodvect_uv = np.cross(u,v)

# Visualizo el resultado

print('prodvect_uv =', prodvect_uv);

prodvect_uv = [ -90 -405  -47]


In [ ]:
# b). Defino los vectores u y v

u = np.array([1,0,0])

v = np.array([0,1,0])

# Hallo el producto vectorial (producto cruz) de u y v, usando el comando np.cross

prodvect_uv = np.cross(u,v)

# Visualizo el resultado

print('prodvect_uv =', prodvect_uv);

prodvect_uv = [0 0 1]


# **Problemas de práctica**

**Ejercicio 1:** Considere los siguientes vectores:

\begin{equation*}
 \vec{u} = (1, -2, 40),
 \hspace{1cm}
 \vec{v} = (0, -78, 94),
\hspace{1cm}
 \vec{w} = (-2, 5, 20),
 \hspace{1cm}
 \vec{z} = (-6, 1, 1).
\end{equation*}

Realice las operaciones indicadas en cada caso:

a). $(2\vec{u} + \vec{v})\cdot (\vec{w}-5\vec{z}).$

b). $\vec{u} \times \vec{w} + 2(\vec{v}\times \vec{z}) -\vec{v}.$

c). $\|\vec{u} + 4\vec{w}\|\vec{u} + \vec{z}.$

d). $\|\vec{u} \times \vec{w}\|- \|\vec{w} \times \vec{z}\|. $


In [ ]:
# Empiece aquí su código



**Ejercicio 2:** Halle el ángulo entre los vectores y basado en esto determine si son ortogonales.

a). $\vec{u} = (2, 9), \ \vec{v} = (-3,7).$

b). $\vec{u} = (1,-5, 2), \ \vec{v} = (-2,0,1).$

In [ ]:
# Empiece aquí su código



**Ejercicio 3:** Sean $\vec{u} = (7, 10, -2)$ y $\vec{v} = (2, −4, 1)$. Hallar dos vectores $\vec{p}$ y $\vec{h}$ tales que $\vec{v} = \vec{p} + \vec{h}$ con $\vec{p}$ paralelo a $\vec{u}$ y $\vec{h}$ ortogonal a $\vec{u}$.

In [ ]:
# Empiece aquí su código

